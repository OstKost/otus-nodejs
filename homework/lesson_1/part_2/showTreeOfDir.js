import fsp from "node:fs/promises";
import path from "path";

const IGNORE_DIRS = [".git", "node_modules", ".idea"];

export const checkDirectory = async (_path) => {
  try {
    const stat = await fsp.stat(_path);
    if (!stat.isDirectory()) throw new Error(`${_path} is not a directory`);
    return true;
  } catch (e) {
    return false;
  }
};

export const getStructure = async (dirPath, maxDepth) => {
  const tree = [];
  if (!maxDepth) return null;
  const dirInfo = await fsp.readdir(dirPath);
  for await (const name of dirInfo) {
    if (IGNORE_DIRS.includes(name)) continue;
    const stat = await fsp.stat(path.join(dirPath, name));
    const isDirectory = stat.isDirectory();
    const children = isDirectory ? await getStructure(path.join(dirPath, name), maxDepth - 1) : null;
    tree.push({
      name,
      isDirectory,
      children
    });
  }
  return tree;
};

export const buildTree = async (structure, title) => {
  let tree = "";
  tree += path.resolve(title);
  const renderElement = (count, element) => Array(count).fill(element).join("");
  const buildLevel = (files = [], depth = 0, isLast = false) => {
    if (!files) return;
    files.forEach((file, index) => {
      const isLastItem = files.length - 1 === index;
      let prefix = "  ";
      const vLinesLimit = isLast ? 1 : 0;
      const renderCount = (depth - vLinesLimit < 0) ? 0 : (depth - vLinesLimit);
      prefix += renderElement(renderCount, "│   ");
      if (isLast) prefix += renderElement(1, "    ");
      prefix += isLastItem ? "└──" : "├──";

      const counterChildren = Array.isArray(file.children) ? `(${file.children.length})` : "";
      const level = `${prefix} ${file.name} ${counterChildren}`.trim() + "\n";
      tree += level;
      if (file.isDirectory) buildLevel(file.children, depth + 1, isLastItem);
    });
  };

  buildLevel(structure);

  return tree;
};
