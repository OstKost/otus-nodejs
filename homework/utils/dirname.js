import path from "path";
import { fileURLToPath } from "url";

const dirname = path.dirname(fileURLToPath(import.meta.url));

export const root = path.join(dirname, '..', '..');