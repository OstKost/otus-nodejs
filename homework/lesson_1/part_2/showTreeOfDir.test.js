import { buildTree, checkDirectory, getStructure } from "./showTreeOfDir.js";
import { root } from "../../utils/dirname.js";
import path from "path";


describe("Renders correct filesystem tree", () => {
  const rootFilePath = path.join(root, "homework", "lesson_1", "part_2", "__fixtures__", "app");

  // checkDirectory
  it("should return false when non exist", async () => {
    const filePath = "any-non-exist";
    const stats = await checkDirectory(filePath);
    expect(stats).toBeFalsy();
  });

  it("should return false if not directory", async () => {
    // const filePath = path.join(root, "homework", "lesson_1", "part_2", "index.js");
    const filePath = rootFilePath + "/dir1/file1.txt";
    const stats = await checkDirectory(filePath);
    expect(stats).not.toBeTruthy();
  });

  it("should return true if initial path exists and is directory", async () => {
    // const filePath = path.join(root, "homework", "lesson_1", "part_2");
    const stats = await checkDirectory(rootFilePath);
    expect(stats).toBeTruthy();
  });

  // getStructure
  it("should be null without maxDepth", async () => {
    // const filePath = path.join(root, "homework");
    const result = await getStructure(rootFilePath);
    expect(result).toBeNull();
  });

  it("should return dir structure with 1 depth", async () => {
    // const filePath = path.join(root, "homework");
    const structure = await getStructure(rootFilePath, 1);
    expect(structure).toMatchSnapshot();
  });

  it("should return dir structure with 2 depth", async () => {
    // const filePath = path.join(root, "homework");
    const structure = await getStructure(rootFilePath, 2);
    expect(structure).toMatchSnapshot();
  });

  it("should return dir structure with depth 3", async () => {
    // const filePath = path.join(root, "homework");
    const structure = await getStructure(rootFilePath, 3);
    expect(structure).toMatchSnapshot();
  });

  // buildTree
  it("should build text from structure - depth 0", async () => {
    // const filePath = path.join(root, "homework");
    const structure = await getStructure(rootFilePath, 0);
    const tree = await buildTree(structure, rootFilePath);
    expect(tree).toMatchSnapshot();
  });

  it("should build text from structure - depth 2", async () => {
    // const filePath = path.join(root, "homework");
    const structure = await getStructure(rootFilePath, 2);
    const tree = await buildTree(structure, rootFilePath);
    expect(tree).toMatchSnapshot();
  });

  it("should build text from structure - depth 3", async () => {
    // const filePath = path.join(root, "homework");
    const structure = await getStructure(rootFilePath, 3);
    const tree = await buildTree(structure, rootFilePath);
    expect(tree).toMatchSnapshot();
  });
});