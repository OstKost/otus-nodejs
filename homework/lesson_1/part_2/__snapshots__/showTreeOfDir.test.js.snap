// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correct filesystem tree should build text from structure - depth 0 1`] = `"/Users/ostkost/Projects/otus-nodejs/homework/lesson_1/part_2/__fixtures__/app"`;

exports[`Renders correct filesystem tree should build text from structure - depth 2 1`] = `
"/Users/ostkost/Projects/otus-nodejs/homework/lesson_1/part_2/__fixtures__/app├── dir1 (3)
│   ├── dir11
│   ├── dir12
│   └── file1.txt
├── dir2 (2)
│   ├── dir21
│   └── dir22
└── dir3 (1)
└── dir31
"
`;

exports[`Renders correct filesystem tree should build text from structure - depth 3 1`] = `
"/Users/ostkost/Projects/otus-nodejs/homework/lesson_1/part_2/__fixtures__/app├── dir1 (3)
│   ├── dir11 (2)
│   │   ├── file111.txt
│   │   └── file112.txt
│   ├── dir12 (0)
│   └── file1.txt
├── dir2 (2)
│   ├── dir21 (1)
│   │   └── file211.txt
│   └── dir22 (1)
│       └── file221.txt
└── dir3 (1)
└── dir31 (0)
"
`;

exports[`Renders correct filesystem tree should return dir structure with 1 depth 1`] = `
[
  {
    "children": null,
    "isDirectory": true,
    "name": "dir1",
  },
  {
    "children": null,
    "isDirectory": true,
    "name": "dir2",
  },
  {
    "children": null,
    "isDirectory": true,
    "name": "dir3",
  },
]
`;

exports[`Renders correct filesystem tree should return dir structure with 2 depth 1`] = `
[
  {
    "children": [
      {
        "children": null,
        "isDirectory": true,
        "name": "dir11",
      },
      {
        "children": null,
        "isDirectory": true,
        "name": "dir12",
      },
      {
        "children": null,
        "isDirectory": false,
        "name": "file1.txt",
      },
    ],
    "isDirectory": true,
    "name": "dir1",
  },
  {
    "children": [
      {
        "children": null,
        "isDirectory": true,
        "name": "dir21",
      },
      {
        "children": null,
        "isDirectory": true,
        "name": "dir22",
      },
    ],
    "isDirectory": true,
    "name": "dir2",
  },
  {
    "children": [
      {
        "children": null,
        "isDirectory": true,
        "name": "dir31",
      },
    ],
    "isDirectory": true,
    "name": "dir3",
  },
]
`;

exports[`Renders correct filesystem tree should return dir structure with depth 3 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": null,
            "isDirectory": false,
            "name": "file111.txt",
          },
          {
            "children": null,
            "isDirectory": false,
            "name": "file112.txt",
          },
        ],
        "isDirectory": true,
        "name": "dir11",
      },
      {
        "children": [],
        "isDirectory": true,
        "name": "dir12",
      },
      {
        "children": null,
        "isDirectory": false,
        "name": "file1.txt",
      },
    ],
    "isDirectory": true,
    "name": "dir1",
  },
  {
    "children": [
      {
        "children": [
          {
            "children": null,
            "isDirectory": false,
            "name": "file211.txt",
          },
        ],
        "isDirectory": true,
        "name": "dir21",
      },
      {
        "children": [
          {
            "children": null,
            "isDirectory": false,
            "name": "file221.txt",
          },
        ],
        "isDirectory": true,
        "name": "dir22",
      },
    ],
    "isDirectory": true,
    "name": "dir2",
  },
  {
    "children": [
      {
        "children": [],
        "isDirectory": true,
        "name": "dir31",
      },
    ],
    "isDirectory": true,
    "name": "dir3",
  },
]
`;
